RecoEgamma/EgammaElectronProducers/plugins/GsfElectronProducer.cc

```diff
@@ -116,6 +152,9 @@ class GsfElectronProducer : public edm::stream::EDProducer<edm::GlobalCache<GsfE
   const bool useGsfPfRecTracks_;
 
   const bool resetMvaValuesUsingPFCandidates_;
+
+  bool dnnPFidEnabled_;
+  std::vector<tensorflow::Session*> elePFid_tfSessions;
```

Comment:
When the TF session is owned by the producer, when running in multithreaded mode in production, each producer will load it's own copy of the models. This can quickly explode the memory usage.
It would be preferrable to load the sessions once in a global cache that is shared across all the producer copies in each thread.
See e.g. here: https://github.com/cms-sw/cmssw/blob/6d2f66057131baacc2fcbdd203588c41c885b42c/RecoMET/METPUSubtraction/plugins/DeepMETProducer.cc#L15

Reply:
In the latest commit I moved the TF:sessions inside the DNNEstimator classes, that are handled inside the global cache both for electrons and photons. 

I haven't been able to define them as std::atomic, thought. The problem is that I need to load a variable number of models and I cannot make atomic the vector of Session*/GraphDefs* pointers.  What do you suggest in this case? Thanks

Reply:
Sorry, I was not sufficiently precise here. If you check the other examples of Session and GraphDef in CMSSW, you should do as is done there. Session is owned by the EDProducer, the GraphDef by the Cache.